inst/private/python_ipc_popen2.m

```diff
@@ -74,9 +93,13 @@
       error('popen2() failed');
     end
 
+    then = @(f, g) g (f ());
     % this will be called when we `clear cleanup`: avoids dangling file handles
-    cleanup = onCleanup (@() python_ipc_popen2_reset (fin, fout, pid));
-
+    cleanup = onCleanup (@() ...
+                          then (@(varagin) ...
+                                 [is_valid_file_id(fin) && fclose(fin), ...
+                                  is_valid_file_id(fout) && fclose(fout)], ...
+                                @(varargin) waitpid(pid)));
```

Comment:
how big a crime is closing here **without** calling `waitpid`?  Then we would not need the `then`...

Note: in the explicit reset case (and a glorious future where `pkg unload` has a hook), we will have waited on `pid` earlier and all of these various will be empty.

Reply:
If we don't `waitpid`, we'll be left with a zombie process `python3 <defunct>`, because `fclose (fin)` will terminate the child process (i.e. `python3`). On the other hand, if we `waitpid` before `fclose (fin)`, it will block, because `waitpid` will wait for the child process (i.e. `python3`) to terminate.

Reply:
thanks!